pinot-common/src/main/java/org/apache/pinot/common/utils/helix/LeadControllerUtils.java

```diff
@@ -0,0 +1,97 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.common.utils.helix;
+
+import java.util.Map;
+import java.util.Set;
+import org.apache.helix.model.ExternalView;
+import org.apache.helix.model.MasterSlaveSMD;
+import org.apache.pinot.common.utils.HashUtils;
+import org.apache.pinot.common.utils.StringUtil;
+
+
+public class LeadControllerUtils {
+
+  /**
+   * Gets hash code for table using murmur2 function, ignores the most significant bit.
+   * Note: This method CANNOT be changed when lead controller resource is enabled.
+   * Otherwise it will assign different controller for the same table, which will mess up the controller periodic tasks and realtime segment completion.
+   * @param rawTableName table name
+   * @return hash code ignoring the most significant bit.
+   */
+  private static int getHashCodeForTable(String rawTableName) {
+    return (HashUtils.murmur2(StringUtil.encodeUtf8(rawTableName)) & 0x7fffffff);
+  }
+
+  /**
+   * Given a raw table name and number of partitions, returns the partition id in lead controller resource.
+   * @param rawTableName raw table name
+   * @param numPartitions number of partitions
+   * @return partition id in lead controller resource.
+   */
+  public static int getPartitionIdForTable(String rawTableName, int numPartitions) {
+    return getHashCodeForTable(rawTableName) % numPartitions;
+  }
+
+  /**
+   * Gets lead controller for table from lead controller resource.
+   * If the resource is disabled or no controller registered as participant, there is no instance in "MASTER" state.
+   * @param leadControllerResourceExternalView external view of lead controller resource
+   * @param rawTableName table name without type
+   * @return leader of partition, null if not found or resource is disabled.
```

Comment:
Please document the format that this return string is in. I believe it is the controller instance name, right?

Reply:
Yes, it's controller partition id like `Controller_localhost_9000`.